src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1016,12 +1174,30 @@ InvalidateDistRelationCacheCallback(Datum argument, Oid relationId)
 		distShardLogicalRelidIndexId = InvalidOid;
 		distShardShardidIndexId = InvalidOid;
 		distShardPlacementShardidIndexId = InvalidOid;
+		distNodeRelationId = InvalidOid;
 		extraDataContainerFuncId = InvalidOid;
 	}
 }
 
 
 /*
+ * InvalidateNodeRelationCacheCallback destroys the WorkerNodeHash when
+ * any change happens on pg_dist_node table. It also set WorkerNodeHash to
+ * NULL, which allows consequent accesses to the hash read from the
+ * pg_dist_node from scratch.
+ */
+static void
+InvalidateNodeRelationCacheCallback(Datum argument, Oid relationId)
+{
+	if (WorkerNodeHash != NULL && relationId == DistNodeRelationId())
+	{
+		hash_destroy(WorkerNodeHash);
```

Comment:
This should move into the initialization function.

https://github.com/citusdata/citus-mx/blob/masterless/main/src/backend/distributed/utils/metadata_cache.c#L1179